### PR TITLE
Renamed DNSArecord to DNSRecord and included DNS C-Names

### PR DESCRIPTION
**Pull request by DdenBraver**
Updated DSC Resource to include DNS C-names.

### CODE REVIEW

**Comment by HemantMahawar** [comment]
@DdenBraver @iainbrighton  @KarolKaczmarek This is a breaking change. On PowerShell gallery, xDnsServer module has been downloaded over **1350** times. Not sure how many of them are using this resource from the module.
Can we gather some data that it is okay to introduce a breaking change?

**Comment by iainbrighton** [comment]
I know that a single resource makes sense. However, could we add the renamed resource whilst leaving the DNSArecord intact and add a deprecated note in the readme? Have we done this with any other resources?

This is essentially forking that resource. Whilst it might increase the work in the short-term, it'll probably be worth it in the long run. After all there are more than just A and CNAME records!

**Comment by DdenBraver** [comment]
@HemantMahawar I fully agree with @iainbrighton. Please let me know if you would like me to add the DNSArecord back to this pull request.
I already have it on my to-do list to update the module for other dns record types.

**Comment by iainbrighton** [comment]
@DdenBraver I now need to be able to create multiple A records as well as CNAME records. What is happening with this pull request? I wouldn't want to duplicate any of this work if this can be merged /cc @TravisEz13 @HemantMahawar

**Comment by TravisEz13** [comment]
@DdenBraver and @iainbrighton,
I'm conflicted.  We shouldn't have two resources that configure the same thing, but breaking compatibility is a bad thing too.

How about this compromise.  Create both resources as suggested, but add a warning to the old one that is has been replaced by the new resource.  File an issue that it should eventually completely remove and reference the issue in the warning.

@BrucePay @HemantMahawar   What do you think?

**Comment by iainbrighton** [comment]
Could we not change the xDnsARecord implementation to a composite resource that simply calls the new xDnsRecord? That way we won't break backwards compatibility and still only have one code base to maintain?

**Comment by DdenBraver** [comment]
I am not 100% certain if adding the xDNSARecord as a composite resource inside the module is the best way to go forward, as this requires additional changes to files that are otherwise un-touched.

I can just copy the xDNSRecord code over to xDNSARecord and add a write-warning with a message like: "DSC Resource xDNSARecord has been replaced by xDNSRecord".
Or I can copy back the old xDNSARecord directory and add a warning message.

@TravisEz13 @iainbrighton @HemantMahawar Please let me know what you think is best.

**Comment by HemantMahawar** [comment]
@DdenBraver @iainbrighton @TravisEz13 ... Here are my thoughts
I think, having a single resource for CName and ARecord makes sense. We can do the following 2 things to existing DNSARecord resource
- Write a warning, as suggested above, that it is deprecated and replaced by DNSRecord
-  **IF we still want to invest & add to DNSARecord resource**, the Get/Set/Test-TargetResource of that resource can call the Get/Set/Test-TargetResource of DNSRecord resource.

I personally, would leave DNSARecord as is with addition of a warning message and invest only in DNSRecord going forward. That way we can move people over to DNSRecord resource. Also it might be useful to add a warning on module load, so it shows up on calls to Get-DscResource ...

File: DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.schema.mof [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[ClassVersion("1.0.0.0"), FriendlyName("xDnsRecord")]
+class MSFT_xDnsRecord : OMI_BaseResource
+{
+    [Key] string Name;
+    [Key] string Zone;
+    [Required,ValueMap{"A-record", "C-name"},Values{"A-record", "C-name"}] string Type;
```
Comment (HemantMahawar) [inline DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.schema.mof]: Should it be CName and ARecord instead? That's what I see online and in powershell cmdlets

File: DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.psm1 [hunk 2]
```diff
@@ -46,12 +60,25 @@ function Set-TargetResource
...
         $Target
     )
-    
-    Write-Verbose "Creating for DNS $Target in $Zone"
-    Add-DnsServerResourceRecordA -IPv4Address $Target -Name $Name -ZoneName $Zone -ComputerName "localhost" 
+
+    if ($Type -eq "A-record")
+    {
+        Write-Verbose "Creating $Type for DNS $Target in $Zone"
+        Add-DnsServerResourceRecordA -IPv4Address $Target -Name $Name -ZoneName $Zone
```
Comment (HemantMahawar) [inline DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.psm1]: Would it be better to use [Add-DnsServerResourceRecord](https://technet.microsoft.com/en-us/library/jj649925.aspx) cmdlet instead along with splatting so adding more types later will not bloat the code. Something like
```powershell
$parameters = @{IPv4Address = $Target; Name=$Name; ZoneName=$Zone} 
```
and in the if/else decide to add -A or -CName switch parameters as
```powershell 
if($type -eq 'CName'){$parameters.Add('CNAME',$true) } 
```
and outside of if/else splat to [Add-DnsServerResourceRecord](https://technet.microsoft.com/en-us/library/jj649925.aspx)  cmdlet as 
```powershell 
Add-DnsServerResourceRecord @parameters
```

File: README.md [hunk 3]
```diff
@@ -30,10 +30,12 @@ Secondary zones allow client machine in primary DNS zones to do DNS resolution o
 Values include: { None | Any | Named | Specific }
 * **SecondaryServer**: IP address or DNS name of DNS servers where zone information can be transfered.
 
-### xDnsARecord
+### xDnsRecord
 * **Name**: Name of the DNS zone
 * **Zone**: The name of the zone to create the record in
 * **Target**: The IP address of the A record
```
Comment (HemantMahawar) [inline README.md]: We should clarify that the Target takes **ONLY IPv4 addresses** in current releas

**Comment by iainbrighton** [comment]
@DdenBraver @HemantMahawar  @TravisEz13 Agreed; putting the original xDNSARecord resource back untouched with a warning will be the safest and easiest solution.

The composite resource suggestion would ensure xDNSARecord resource functionality would continue to work without breaking anything _and_ inherit any future improvements in the xDNSRecord resource without any further changes, tests or modifications! However, it's still a change.

**Comment by DdenBraver** [comment]
@iainbrighton @HemantMahawar Thank you for all the feedback, and comments on the code.
I will do my best to edit and test the code this weekend.

**Comment by TravisEz13** [comment]
For your test failure, please see these issues:
[DscResource.Tests #22](https://github.com/PowerShell/DscResource.Tests/issues/22)
[xDscDesigner PR #36](https://github.com/PowerShell/xDSCResourceDesigner/pull/36/files)

Which should help resolve the issue.

**Comment by DdenBraver** [comment]
@TravisEz13 Thank you, I somehow always struggle with either UTF8 or Unicode errors.

@HemantMahawarI updated the resource using your feedback, however do you have any hint where I can place the warning message so that it will show on module load? Currently I just added it to the get-targetresource so that it will show every time when used.

@iainbrighton I added the "Ensure" parameter to the DNSRecord resource now as well.

**Comment by TravisEz13** [comment]
It would be better to add the warning to Test-targetresource.  Test target resource is always called during consistency.  Get-targetresource is rarely called.  Module load is not a good place to put error or warnings.

**Comment by DdenBraver** [comment]
@TravisEz13 In this case, I placed it into the Get-TargetResource since in this case it is called by the Test-TargetResource.
If you still think it would be best to move it, please let me know.

**Comment by iainbrighton** [comment]
Cool! Does it support creating multiple A records for round-robin operation? I think I logged that as a separate issue?

**Comment by DdenBraver** [comment]
@iainbrighton Just did a few tests, and fixed the multiple A records support.

**Comment by iainbrighton** [comment]
You star :)

**Comment by TravisEz13** [comment]
@DdenBraver   If get is always called by test, putting it in Get is fine.

Re:
>@TravisEz13 In this case, I placed it into the Get-TargetResource since it is called by the Test-   TargetResource.
>If you still think it would be best to move it, please let me know.

**Comment by TravisEz13** [comment]
@DdenBraver Re: UTF-8/Unicode error
I have pull requests completed and outstanding to clarify the common problems I've seen.  Hopefully this will help.

File: DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.schema.mof [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[ClassVersion("1.0.0.0"), FriendlyName("xDnsRecord")]
+class MSFT_xDnsRecord : OMI_BaseResource
+{
+    [Key] string Name;
```
Comment (TravisEz13) [inline DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.schema.mof]: Please try to use descriptions in the schema.

File: Tests/xDnsRecord.Tests.ps1 [hunk 5]
```diff
@@ -11,22 +11,24 @@ Set-StrictMode -Version latest
...
+            Type = "ARecord"
             Ensure = "Present"
         }
         $testAbsentParams = @{
             Name = $testPresentParams.Name
             Zone = $testPresentParams.Zone
             Target = $testPresentParams.Target
+            Type = $testPresentParams.Type
             Ensure = "Absent"
         }
```
Comment (TravisEz13) [inline Tests/xDnsRecord.Tests.ps1]: Should have a test that Get-targetresource correctly returns the current type of the record.

**Comment by TravisEz13** [comment]
Thanks for the contribution!
Please review additional comments for future contribution Guidance.

**Pull request merged**